### PR TITLE
#342 - Add caching lazy MinkowskiSum

### LINKED ISSUE #342: Add caching lazy MinkowskiSum
In `Reachability.jl` we can build lazy sums in a loop. In each loop iteration we query the support vector of the sum and add one more element. It would be more efficient if the sum stored the support vector result for most summands and only computed the support vector of the remaining set(s).

Example:
```julia
sum = CSum([X1, X2, X3])
σ(d, sum)
# in the background, `sum` computes `r1 = σ(d, X1) + σ(d, X2) + σ(d, X3)`
# and stores ``[X1, X2, X3] => [(d, r1)]`
sum = sum + X4
# in the background, `sum` stores `[X1, X2, X3] => [(d, r1)], [X4] => []`
σ(d, sum)
# in the background, `sum` computes `r2 = r1 + σ(d, X4)`
# and stores `[X1, X2, X3, X4] => [(d, r2)]`
```

It could be helpful to have #269 first.

### PR DESCRIPTION
**Pull request by schillic**
Closes #342.

### CODE REVIEW

**Comment by schillic** [comment]
```julia
julia> using LazySets

julia> cms = CacheMinkowskiSum(5);

julia> X1 = BallInf(ones(3), 3.);

julia> X2 = BallInf(ones(3), 5.);

julia> X3 = Ball1(ones(3), 5.);

julia> d = ones(3);

julia> a = array(cms);

julia> 

julia> push!(a, X1);

julia> @time σ(d, cms);
  0.000027 seconds (25 allocations: 1.500 KiB)

julia> @time σ(d, cms);
  0.000012 seconds (12 allocations: 608 bytes)

julia> @time σ(d, cms);
  0.000011 seconds (12 allocations: 608 bytes)

julia> 

julia> push!(a, X2);

julia> @time σ(d, cms);
  0.000019 seconds (27 allocations: 1.656 KiB)

julia> @time σ(d, cms);
  0.000012 seconds (12 allocations: 608 bytes)

julia> @time σ(d, cms);
  0.000010 seconds (12 allocations: 608 bytes)

julia> 

julia> for i = 1 : 1000
           push!(a, X3);
       end

julia> @time σ(d, cms);
  0.000160 seconds (3.02 k allocations: 329.469 KiB)

julia> @time σ(d, cms);
  0.000023 seconds (13 allocations: 624 bytes)

julia> @time σ(d, cms);
  0.000015 seconds (13 allocations: 624 bytes)

julia> 

julia> for i = 1 : 1000
           push!(a, X3);
       end

julia> @time σ(d, cms);
  0.000159 seconds (3.03 k allocations: 329.484 KiB)

julia> @time σ(d, cms);
  0.000028 seconds (13 allocations: 624 bytes)

julia> @time σ(d, cms);
  0.000011 seconds (13 allocations: 624 bytes)
```

```julia
julia> msa = MinkowskiSumArray(5);

julia> a2 = array(msa);

julia> push!(a2, X1);

julia> push!(a2, X2);

julia> for i = 1 : 1000
           push!(a2, X3);
       end

julia> @time σ(d, msa);
  0.000162 seconds (3.02 k allocations: 329.547 KiB)

julia> @time σ(d, msa);
  0.000150 seconds (3.02 k allocations: 329.547 KiB)

julia> @time σ(d, msa);
  0.000152 seconds (3.02 k allocations: 329.547 KiB)

julia> for i = 1 : 1000
           push!(a2, X3);
       end

julia> @time σ(d, msa);
  0.000274 seconds (6.02 k allocations: 657.672 KiB)

julia> @time σ(d, msa);
  0.000257 seconds (6.02 k allocations: 657.672 KiB)

julia> @time σ(d, msa);
  0.000989 seconds (6.02 k allocations: 657.672 KiB)

julia> @time σ(d, msa);
  0.000897 seconds (6.02 k allocations: 657.672 KiB)
```

**Comment by schillic** [comment]
Possible future work: Currently the sum stores all sets, which takes a lot of memory. In applications where we always query the same support vector(s) and then add a new set in a loop, we could also provide a variant that only stores the newest sets.

File: src/MinkowskiSum.jl [hunk 1]
```diff
@@ -201,8 +201,153 @@ The support vector in the given direction.
...
+
+# constructor for an empty sum with optional size hint and numeric type
+function CacheMinkowskiSum(n::Int=0, N::Type=Float64)::CacheMinkowskiSum
+    arr = Vector{LazySet{N}}(0)
+    sizehint!(arr, n)
+    return CacheMinkowskiSum(arr)
+end
+
+# ZeroSet is the neutral element for CacheMinkowskiSum
+@neutral(CacheMinkowskiSum, ZeroSet)
```
Comment (mforets) [inline src/MinkowskiSum.jl]: 🏆

File: src/MinkowskiSum.jl [hunk 2]
```diff
@@ -201,8 +201,153 @@ The support vector in the given direction.
...
+Constructors:
+
+- `CacheMinkowskiSum(array::Vector{<:LazySet})` -- default constructor
+
+- `CacheMinkowskiSum([n]::Int=0, [N]::Type=Float64)`
+  -- constructor for an empty sum with optional size hint and numeric type
+"""
+struct CacheMinkowskiSum{N<:Real, S<:LazySet{N}} <: LazySet{N}
+    array::Vector{S}
+    cache::Dict{AbstractVector{N}, Tuple{Int, AbstractVector{N}}}
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/MinkowskiSum.jl]: That is true, but the directions are always constructed from the same sources at the moment, which is why we get the same numbers everytime.

File: src/MinkowskiSum.jl [hunk 3]
```diff
@@ -201,8 +201,153 @@ The support vector in the given direction.
...
+"""
+    CacheMinkowskiSum{N<:Real, S<:LazySet{N}} <: LazySet{N}
+
+Type that represents the Minkowski sum of a finite number of convex sets.
+Support vector queries are cached.
+
+### Fields
+
+- `array` -- array of convex sets
+- `cache` -- cache of support vector query results
```
[the remark this answers is not in the document]
Comment (schillic) [reply src/MinkowskiSum.jl]: added :+1:

**Pull request merged**